ui/src/app/controllers/components/ControllerHeaderForms/AddController/AddController.test.tsx

```diff
@@ -0,0 +1,58 @@
+import { mount } from "enzyme";
+import { Provider } from "react-redux";
+import { MemoryRouter } from "react-router-dom";
+import configureStore from "redux-mock-store";
+
+import AddController from "./AddController";
+
+import type { RootState } from "app/store/root/types";
+import {
+  configState as configStateFactory,
+  rootState as rootStateFactory,
+} from "testing/factories";
+
+const mockStore = configureStore();
+
+describe("AddController", () => {
+  let state: RootState;
+
+  beforeEach(() => {
+    state = rootStateFactory({
+      config: configStateFactory({
+        items: [
+          { name: "maas_url", value: "http://1.2.3.4/MAAS" },
+          { name: "rpc_shared_secret", value: "veryverysecret" },
+        ],
+      }),
+    });
+  });
```

Comment:
This is a minor thing, but I noticed you often use this pattern. It seems unnecessary (unless there's something that I'm missing, please explain).

You could get rid of reassignement of state in `beforeEach` and nesting all tests in `describe` by returning a new state object in each test (like you do with `store`). This would simplify this test and make easier to reason about.

https://kentcdodds.com/blog/avoid-nesting-when-youre-testing

Reply:
The pattern we've adopted is to set up the root state in beforeEach that is common to all the tests (which can be quite complex) and then override the parts of state that is required for a single test.

For a simple example in the App tests we have:

```
describe("App", () => {
  let state: RootState;

  beforeEach(() => {
    state = rootStateFactory({
      config: configStateFactory({
        items: [{ name: "completed_intro", value: true }],
      }),
    });
  });

  it("renders routes if logged in", () => {
    state.status.connected = true;
    state.status.authenticated = true;
    const store = mockStore(state);
    const wrapper = mount(
     ...
  });
```

This way we don't have to override the completed_intro config in every test, but it's also explicit what state is required for a single test.

Reply:
Thanks, good suggestion. I've updated it so it looks like this now:

<img width="1307" alt="Screen Shot 2022-01-12 at 9 46 08 am" src="https://user-images.githubusercontent.com/361637/149033452-6f6c9a11-1115-4241-b8dc-2a76bc2f4ddf.png">